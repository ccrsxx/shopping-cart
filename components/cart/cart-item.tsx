import Link from 'next/link';
import { motion } from 'framer-motion';
import { useShoppingCart } from '@lib/context/shopping-cart';
import { formatCurrency } from '@lib/currency';
import { setTransition } from '@lib/transition';
import { MdAdd, MdRemove, MdDelete } from '@assets/icons';
import { ImageLoader } from '@components/ui/image-loader';
import { Button } from '@components/ui/button';

type CartItemProps = {
  id: number;
  index: number;
  title: string;
  image: string;
  price: number;
  quantity: number;
  toggleCart: () => void;
};

export function CartItem({
  id,
  index,
  title,
  image,
  price,
  quantity,
  toggleCart
}: CartItemProps): JSX.Element {
  const { deleteProduct, handleProductQuantity } = useShoppingCart();

  const toProduct = `/product/${id}`;

  return (
    <motion.li
      {...setTransition({
        layout: true,
        delayIn: index * 0.02,
        direction: 'right',
        durationOut: 0.25
      })}
    >
      <div className='flex rounded-lg border border-border-primary'>
        <Link href={toProduct}>
          <a
            className='tab hidden rounded-r-none transition hover:brightness-90 
                       focus-visible:brightness-90 md:block'
            onClick={toggleCart}
          >
            <ImageLoader
              divStyle='h-[108px] w-[108px] rounded-l-lg shrink-0 shrink-0 bg-white'
              imageStyle='!p-4'
              src={image}
              alt={title}
            />
          </a>
        </Link>
        <div className='flex w-full flex-col justify-between py-2 px-4'>
          <div>
            <Link href={toProduct}>
              <a
                className='tab overflow-hidden text-ellipsis font-medium text-white/90
                           transition [display:-webkit-box] [-webkit-line-clamp:1]
                           [-webkit-box-orient:vertical] hover:text-white'
                onClick={toggleCart}
              >
                {title}
              </a>
            </Link>
            <p className='font-bold'>{formatCurrency(price)}</p>
          </div>
          <div className='flex flex-wrap justify-between text-sm font-light'>
            <div className='flex gap-1'>
              <p>Quantity: </p>
              <input
                className='max-w-[60px] rounded-lg bg-background px-1 transition hover:ring-2 
                           hover:ring-accent focus:outline-none focus:ring-2 focus:ring-accent'
                type='number'
                min={1}
                max={10_000}
                value={quantity}
                onChange={handleProductQuantity(id)}
              />
            </div>
            <p>
              Total:{' '}
              <span className='font-semibold'>
                {formatCurrency(price * quantity)}
              </span>
            </p>
          </div>
        </div>
        <div className='grid border-l border-border-primary text-sm inner:rounded-none'>
          <Button
            className='!rounded-tr-lg disabled:text-gray-600 disabled:!brightness-100'
            ariaLabel='Increase'
            Icon={MdAdd}
            onClick={handleProductQuantity(id, 'increment')}
            disabled={quantity >= 10_000}
          />
          <Button
            className='border-y border-border-primary disabled:text-gray-600 disabled:!brightness-100'
            ariaLabel='Decrease'
            Icon={MdRemove}
            onClick={handleProductQuantity(id, 'decrement')}
            disabled={quantity <= 1}
          />
          <Button
            className='!rounded-br-lg'
            ariaLabel='Delete'
            Icon={MdDelete}
            onClick={deleteProduct(id)}
          />
        </div>
      </div>
    </motion.li>
  );
}
