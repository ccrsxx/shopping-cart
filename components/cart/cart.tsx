import { useState, useEffect } from 'react';
import Link from 'next/link';
import { AnimatePresence, motion } from 'framer-motion';
import { MdShoppingCart, MdArrowForward } from '@assets/icons';
import { Button } from '@components/ui/button';
import { useShoppingCart } from '@lib/context/shopping-cart';
import { formatCurrency } from '@lib/currency';
import { setTransition } from '@lib/transition';
import { CartItem } from './cart-item';

export function Cart(): JSX.Element {
  const { cartProducts, currentCart, totalPrice, isMobile, clearCart } =
    useShoppingCart();

  const [totalProducts, setTotalProducts] = useState(0);
  const [isCartOpen, setIsCartOpen] = useState(false);

  useEffect(() => {
    setTotalProducts(cartProducts);
  }, [cartProducts]);

  useEffect(() => {
    if (isMobile && !isCartOpen) document.documentElement.style.overflow = '';

    document[isMobile ? 'documentElement' : 'body'].style.overflowY = isCartOpen
      ? 'hidden'
      : '';
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isCartOpen]);

  const toggleCart = (): void => setIsCartOpen(!isCartOpen);

  const cartLength = currentCart.length;

  return (
    <>
      <Button className='relative !p-2' ariaLabel='Open' onClick={toggleCart}>
        <MdShoppingCart size={24} />
        <span
          className='absolute -right-0.5 -top-0.5 flex h-5 min-w-[20px] items-center
                     justify-center rounded-full bg-main-red p-1 text-xs'
        >
          {totalProducts}
        </span>
      </Button>
      <AnimatePresence exitBeforeEnter>
        {isCartOpen && (
          <div className='fixed inset-0 z-10'>
            <motion.div
              className='fixed h-screen w-screen bg-black/80'
              onClick={toggleCart}
              {...setTransition({ direction: 'none' })}
            />
            <motion.div
              className='fixed top-0 -right-4 h-screen w-[500px] max-w-[90vw]
                         rounded-l-lg bg-background p-8 pr-12'
              {...setTransition({
                direction: 'right',
                distance: 'full',
                durationIn: 0.5
              })}
            >
              <div className='flex h-full flex-col gap-4'>
                <div className='flex justify-between'>
                  <h2 className='text-2xl font-bold'>
                    {cartLength
                      ? `${cartLength} product${cartLength > 1 ? 's' : ''}`
                      : 'No product added'}
                  </h2>
                  <Button
                    className='!justify-end !p-0 text-xl normal-case text-secondary hover:scale-100 hover:!bg-inherit'
                    label='Clear'
                    onClick={clearCart}
                  />
                </div>
                <ul
                  className='relative order-1 -mx-8 -my-1 -mr-6 flex flex-1 flex-col gap-4
                             overflow-y-auto overflow-x-hidden py-1 px-8 pr-6'
                >
                  <AnimatePresence>
                    {currentCart.map((cartProduct, index) => (
                      <CartItem
                        index={index}
                        toggleCart={toggleCart}
                        key={cartProduct.id}
                        {...cartProduct}
                      />
                    ))}
                  </AnimatePresence>
                </ul>
                <div className='flex flex-wrap items-center justify-between md:order-2'>
                  <p className='text-lg'>
                    Total:{' '}
                    <span className='font-bold'>
                      {formatCurrency(totalPrice)}
                    </span>
                  </p>
                  <Link href='/checkout'>
                    <a tabIndex={-1} onClick={toggleCart}>
                      <Button
                        Icon={MdArrowForward}
                        label='Checkout'
                        className='text-xl normal-case hover:!bg-inherit hover:!text-accent'
                        flipped
                      />
                    </a>
                  </Link>
                </div>
              </div>
            </motion.div>
          </div>
        )}
      </AnimatePresence>
    </>
  );
}
