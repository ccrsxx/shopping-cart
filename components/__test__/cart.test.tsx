import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { Cart } from '@components/cart/cart';

export type CartProps = { name: string; price: number; quantity: number };

interface MockProps {
  cartProducts: number;
  currentCart: CartProps[];
  totalPrice: number;
  clearCart: () => void;
}

let mockCurrentCart = [
  { id: 1, name: 'emilia', price: 10, quantity: 2 },
  { id: 2, name: 'rem', price: 9, quantity: 1 },
  { id: 3, name: 'lena', price: 7, quantity: 3 }
];

const mockClearCart = (): void => {
  mockCurrentCart = [];
};

const [mockCartProducts, mockTotalPrice] = mockCurrentCart.reduce(
  ([products, total], { price, quantity }) => [
    products + quantity,
    total + price * quantity
  ],
  [0, 0]
);

jest.mock('@lib/context/shopping-cart', () => ({
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  useShoppingCart: (): MockProps => ({
    cartProducts: mockCartProducts,
    currentCart: mockCurrentCart,
    totalPrice: mockTotalPrice,
    clearCart: mockClearCart
  })
}));

jest.mock('@components/cart/cart-item', () => ({
  CartItem: (): JSX.Element => <div data-testid='cart-item' />
}));

let rerenderComponent: (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>
) => void;

describe('test Cart', () => {
  beforeEach(() => {
    const { rerender } = render(<Cart />);
    rerenderComponent = rerender;
  });

  function getCartItems(): HTMLElement[] {
    return screen.queryAllByTestId('cart-item');
  }

  it('renders correctly', () => {
    const { container } = render(<Cart />);
    expect(container).toMatchSnapshot();
  });

  it('should not render cart items in the beginning', () => {
    const cartItems = getCartItems();

    expect(cartItems).toHaveLength(0);
  });

  it('renders cart items when cart button is clicked', () => {
    const cartButton = screen.getByRole('button');

    userEvent.click(cartButton);

    const cartItems = getCartItems();

    expect(cartItems).toHaveLength(3);
  });

  it('should clear cart items when clear button is clicked', () => {
    const cartButton = screen.getByRole('button');

    userEvent.click(cartButton);

    const beforeClickedCartItems = getCartItems();

    expect(beforeClickedCartItems).toHaveLength(3);

    const clearButton = screen.getByRole('button', {
      name: /clear/i
    });

    userEvent.click(clearButton);

    rerenderComponent(<Cart />);
    rerenderComponent(<Cart />);

    const afterClickedCartItems = getCartItems();

    expect(afterClickedCartItems).toHaveLength(0);
  });
});
