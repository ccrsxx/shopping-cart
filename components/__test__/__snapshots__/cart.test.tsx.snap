// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Cart renders correctly 1`] = `
<div>
  <button
    aria-label="Open"
    class="tab flex select-none items-center justify-center gap-2 rounded-lg bg-transparent py-2 px-3 
         font-medium transition duration-300 hover:text-primary active:scale-90 active:duration-150 
         enabled:hover:bg-accent disabled:cursor-not-allowed disabled:brightness-50 relative !p-2"
    type="button"
  >
    <svg
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49A1.003 1.003 0 0020 4H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
    <span
      class="absolute -right-0.5 -top-0.5 flex h-5 min-w-[20px] items-center justify-center rounded-full bg-main-red p-1 text-xs"
    >
      6
    </span>
  </button>
</div>
`;
